File: junit4-rules/src/test/java/com/vanniktech/junit4rules/DefaultLocaleRuleTest.java
package com.vanniktech.junit4rules;

import java.util.Locale;
import org.junit.FixMethodOrder;
import org.junit.Rule;
import org.junit.Test;

import static java.util.Locale.KOREAN;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.runners.MethodSorters.NAME_ASCENDING;

@FixMethodOrder(NAME_ASCENDING) public final class DefaultLocaleRuleTest {
  private static final Locale DEFAULT;

  static {
    DEFAULT = Locale.getDefault();
  }

  @Rule public final DefaultLocaleRule defaultLocaleRule = new DefaultLocaleRule();

  @Test public void firstShouldKeepDefault() {
    assertEquals(DEFAULT, Locale.getDefault());

    Locale.setDefault(KOREAN); // Mutate the actual locale.
  }

  @Test(expected = UnsupportedOperationException.class) public void secondShouldNotBeAffected() {
    assertNotEquals(KOREAN, Locale.getDefault()); // Mutation from previous test should not affect this one.

    throw new UnsupportedOperationException(); // Let it crash on purpose for the next method.
  }

  @Test public void thirdShouldNotBeAffectedEvenInCaseOfACrash() {
    assertEquals(DEFAULT, Locale.getDefault());
  }
}

File: junit4-android-integration-rules/src/androidTest/java/com/vanniktech/junit4androidintegrationrules/DefaultLocaleRuleTest.kt
package com.vanniktech.junit4androidintegrationrules

import android.annotation.SuppressLint
import androidx.test.core.app.ActivityScenario
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Rule
import org.junit.Test
import java.util.Locale
import java.util.Locale.KOREAN

class DefaultLocaleRuleTest {
  @get:Rule val activityScenario = ActivityScenario.launch(ForceLocaleRuleActivity::class.java)
  @get:Rule val forceLocalRule = DefaultLocaleRule()

  @Test fun firstShouldKeepDefault() {
    activityScenario.onActivity { assertEquals("en_US", it.textView.text) }
    assertEquals(default, Locale.getDefault())

    Locale.setDefault(KOREAN) // Mutate the actual locale.
  }

  @Test(expected = UnsupportedOperationException::class) fun secondShouldNotBeAffected() {
    activityScenario.onActivity { assertEquals("en_US", it.textView.text) }
    assertNotEquals(KOREAN, Locale.getDefault()) // Mutation from previous test should not affect this one.

    throw UnsupportedOperationException() // Let it crash on purpose for the next method.
  }

  @Test fun thirdShouldNotBeAffectedEvenInCaseOfACrash() {
    activityScenario.onActivity { assertEquals("en_US", it.textView.text) }
    assertEquals(default, Locale.getDefault())
  }

  companion object {
    @SuppressLint("ConstantLocale") val default = Locale.getDefault()
  }
}
